### PR TITLE
Create tokens for groups

### LINKED ISSUE #3: Automatic generation of tokens
Automatic generation of tokens 

After creating a group, a button should be available that leads the admin to a form where to choose the number tokens to be generated.This task generates all the tokens using  a secure method, preventing any duplication from existing ones (screenshot provided)

imatge

**Tasks**
- [x] Migration to create table access_tokens with fields: via #8 
  - [x] `belongs_to` `tokens_group` previously created
  - [x] token (string)
  - [x] Num. uses counter (integer)
- [x] Create controller and routes. Access through the previous `tokens_group` controller
- [ ] CRUD operations with these particulars:
  - [ ] Create is generic, just put a number of tokens to be generated. Creation, just creates more tokens, does not remove previous ones. The command in charge of creation should generate a token based on the token generator created in  #10
  - [ ] No edit
  - [ ] Removal as long as the token is not being used.
  - [ ] A button to remove all unused tokens at once
- [x] Prevent removal for `tokens_groups` if tokens exist
- [ ] Add tests

### PR DESCRIPTION
**Pull request by microstudi**
fixes #3 

tests:
- [x] Create tokens command test
- [x] Create tokens form test
- [x] Create tokens system test

### CODE REVIEW

File: app/views/decidim/anonymous_codes/admin/codes/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<div class="card">
+  <div class="card-divider">
+    <h2 class="card-title">
+      <%= t("codes.index.title", scope: "decidim.anonymous_codes.admin", group: translated_attribute(code_group.title)) %>
+      <a class="button tiny button--title new-access" href="<%= code_groups_path(code_group) %>"><%= t("codes.index.back", scope: "decidim.anonymous_codes.admin") %></a>
```
Comment (microstudi) [inline app/views/decidim/anonymous_codes/admin/codes/index.html.erb]:
```suggestion
      <a class="button tiny button--title new-access" href="<%= code_groups_path %>"><%= t("codes.index.back", scope: "decidim.anonymous_codes.admin") %></a>
```

File: config/locales/en.yml [hunk 2]
```diff
@@ -45,6 +46,11 @@ en:
           update:
             invalid: There was a problem updating this access code group
             success: Access code group successfully updated
+        codes:
+          index:
+            back: Back to group
```
Comment (microstudi) [inline config/locales/en.yml]:
```suggestion
            back: Back to groups
```

File: app/forms/decidim/anonymous_codes/admin/tokens_form.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Decidim
+  module AnonymousCodes
+    module Admin
+      class TokensForm < Decidim::Form
+        include TranslatableAttributes
```
Comment (microstudi) [inline app/forms/decidim/anonymous_codes/admin/tokens_form.rb]:
```suggestion
```

File: app/controllers/decidim/anonymous_codes/admin/codes_controller.rb [hunk 4]
```diff
@@ -49,11 +65,11 @@ def export
         private
 
         def tokens
-          AnonymousCodes::Token.for(code_group)
+          AnonymousCodes::Token.where(group: code_group)
         end
 
         def token
-          @token ||= tokens.find(params[:id])
+          tokens.find(params[:id])
```
Comment (microstudi) [inline app/controllers/decidim/anonymous_codes/admin/codes_controller.rb]:
```suggestion
          @token ||= tokens.find(params[:id])
```

File: app/controllers/decidim/anonymous_codes/admin/codes_controller.rb [hunk 5]
```diff
@@ -49,11 +65,11 @@ def export
         private
 
         def tokens
-          AnonymousCodes::Token.for(code_group)
+          AnonymousCodes::Token.where(group: code_group)
```
Comment (microstudi) [inline app/controllers/decidim/anonymous_codes/admin/codes_controller.rb]:
```suggestion
          AnonymousCodes::Token.for(code_group)
```

File: app/forms/decidim/anonymous_codes/admin/tokens_form.rb (hunk 3 above)
Comment (ElviaBth) [reply app/forms/decidim/anonymous_codes/admin/tokens_form.rb]: Done

File: app/controllers/decidim/anonymous_codes/admin/codes_controller.rb (hunk 4 above)
Comment (ElviaBth) [reply app/controllers/decidim/anonymous_codes/admin/codes_controller.rb]: Done

File: app/controllers/decidim/anonymous_codes/admin/codes_controller.rb (hunk 5 above)
Comment (ElviaBth) [reply app/controllers/decidim/anonymous_codes/admin/codes_controller.rb]: Done

File: spec/commands/admin/create_tokens_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        let(:form) { double("FormObject", invalid?: false, num_tokens: 10) }
+        let(:code_group) { double("CodeGroup") }
+        let(:command) { described_class.new(form, code_group) }
+
+        describe "#call" do
+          context "when the form is valid" do
+            before do
+              allow(CreateTokensJob).to receive(:perform_later)
+              command.call
+            end
```
Comment (microstudi) [inline spec/commands/admin/create_tokens_spec.rb]:
```suggestion
        describe "#call" do
          before do
            allow(CreateTokensJob).to receive(:perform_later)
            command.call
          end
          context "when the form is valid" do
```
Let's move on the top so is valid for all examples

File: spec/commands/admin/create_tokens_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+          context "when the form is invalid" do
+            let(:form) { double("FormObject", invalid?: true) }
+
+            it "broadcasts :invalid" do
+              expect(command).to broadcast(:invalid)
+            end
+
+            it "does not queue a background job" do
+              allow(CreateTokensJob).to receive(:perform_later)
+              command.call
```
Comment (microstudi) [inline spec/commands/admin/create_tokens_spec.rb]:
```suggestion
```
Now you don't need this here

File: spec/controller/admin/codes_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+          let!(:token2) { create(:anonymous_codes_token, group: code_group, created_at: 1.day.ago) }
+
+          it "enforces permission to view anonymous code tokens" do
+            expect(controller).to receive(:enforce_permission_to).with(:view, :anonymous_code_token)
+            get :index, params: { code_group_id: code_group.id }
+          end
+
+          it "assigns @tokens with paginated tokens ordered by creation date" do
+            token1
+            token2
```
Comment (microstudi) [inline spec/controller/admin/codes_controller_spec.rb]:
```suggestion
```
This should not be necessary as `token1` and `token2` are created using `let!`

**Review by microstudi — COMMENTED** [review]
Excelent Job Elvia,

left you some simple comments only.

I'd like to add a test for the CreateTokensJob, to check if an generated token is already in the database, let's make sure that it works correctly.

Here's a spec that we could be using (to be placed under `spec/jobs/create_tokens_job_spec.rb`):

```ruby
# frozen_string_literal: true

require "spec_helper"

module Decidim
  module AnonymousCodes
    describe CreateTokensJob do
      subject { described_class.perform_now(code_group, num_tokens) }

      let(:organization) { create(:organization) }
      let(:code_group) { create :anonymous_codes_group }
      let(:num_tokens) { 2 }
      let(:token1) { "TOKEN1" }
      let(:token2) { "TOKEN2" }
      let(:token3) { "TOKEN3" }

      before do
        allow(Decidim::AnonymousCodes).to receive(:token_generator).and_return(token1, token2, token3)
      end

      it "generates 2 tokens" do
        expect { subject }.to change { Token.count }.by(2)
        expect(Token.all.pluck(:token)).to match_array([token1, token2])
      end

      context "when token is repeated" do
        let!(:existing_token) { create(:anonymous_codes_token, token: token2, group: code_group) }

        it "skips the repeated" do
          expect { subject }.to change { Token.count }.by(2)
          expect(Token.all.pluck(:token)).to match_array([token1, token2, token3])
        end
      end
    end
  end
end
```

File: spec/commands/admin/create_tokens_spec.rb (hunk 6 above)
Comment (ElviaBth) [reply spec/commands/admin/create_tokens_spec.rb]: Oh ok. Thanks.

File: spec/commands/admin/create_tokens_spec.rb (hunk 7 above)
Comment (ElviaBth) [reply spec/commands/admin/create_tokens_spec.rb]: Done

File: spec/controller/admin/codes_controller_spec.rb (hunk 8 above)
Comment (ElviaBth) [reply spec/controller/admin/codes_controller_spec.rb]: I see, thanks.

**Pull request merged**